swarm_dial.go

```diff
@@ -158,7 +159,11 @@ func (s *Swarm) Dial(ctx context.Context, p peer.ID) (*Conn, error) {
 		return nil, ErrDialToSelf
 	}
 
-	return s.gatedDialAttempt(ctx, p)
+	conn, err := s.gatedDialAttempt(ctx, p)
+	if s.protec != nil && strings.Contains(err.Error(), "privnet") {
+		log.Warninf("error: %s. Possible missmatched private network key", err.Error())
+	}
+	return conn, err
```

Comment:
You could even have a "root" error type in go-libp2p-interface-pnet, and more specific error types derived from that in the implementation package(s).